### ISSUE TITLE
product search (admin) fails...

### ISSUE THREAD

**Issue by proseLA** [zencart/zencart#3729]
https://github.com/zencart/zencart/blob/3acda696b5ff7b8a2ec7d1aad3b7b25c55dc44bf/admin/category_product_listing.php#L689

i have never understood why the master category is required as part of search.

a user duplicated a product to a new category, and as part of the duplication, ZC did not add a record into the products_to_category table for the master category of the product.  now search fails.

pretty easy to duplicate.  see:

https://www.zen-cart.com/showthread.php?226951-Can-t-search-for-duplicate-after-saving-(Numinix-custom-fields)&p=1370670#post1370670

(Numinix has nothing to do with it.  i was able to reproduce on a vanilla v157 install.)

**Comment by scottcwilson** [comment]
> ... ZC did not add a record into the products_to_category table for the master category of the product. 
to me, that's the fix that's needed. (Or at least part of it.) 

Not sure about the SQL change, but creating a P2C record seems like an obvious need.

**Comment by proseLA** [comment]
> Not sure about the SQL change, but creating a P2C record seems like an obvious need.

the more i think about it, the more i think we do need a SQL change.

why are we preventing store admins from finding their product?  because the data is messed up?  is that a reason to not allow admins to find product?

i am all for fixing the creation of a P2C record, but when store owners create new product types and then code their own copy product module, is it not easy for them to miss the code that creates the P2C for the master category?  what about the data that already exists out in the ZC ecosphere?

in addition, with the exception of active status, why is the search on the storefront different from the search on the admin?  on the storefront one can search for `bug collect` and find item 34; but on the admin we find nothing. 

now, to address the multi word issue, we could do something like i did for customer search here:

https://github.com/zencart/zencart/pull/3079

but i think it would behoove us to have 1 consistent place where we do product search.  perhaps it is a lot of work with refactoring and the like, but it makes sense to me.

finally, as someone who has run across this master category problem before (and addressed it on clients sites), can someone give me a reason why we keep the sql matching the master category in?  if the item's master category does not have a record in the products to categories table, we do not allow admins to find it. why?

**Comment by torvista** [comment]
In my test, duplicating a product from catA to catB:

new product has master category of catA as it uses the same as source product - incorrectly.

new product has p2c category of catB - correct, as per product_copy_confirm.php:
```
        $db->Execute("INSERT INTO " . TABLE_PRODUCTS_TO_CATEGORIES . " (products_id, categories_id)
                      VALUES (" . $dup_products_id . ", " . $categories_id . ")");
```
I think an extra line after this is the easiest fix, no?

**Comment by drbyte** [comment]
I agree that for the search the forced join with P2C is problematic in its current state.

You said that you've "addressed it on clients sites". What did you change? What are the side-effects of the change?

**Comment by torvista** [comment]
>  if the item's master category does not have a record in the products to categories table, we do not allow admins to find it. why?

I can't answer that, but it has flagged up this error, which is good. I've also had p2c/mcat issues in the past and now have a permanent check in place in the admin for products with an invalid mcat.

**Comment by drbyte** [comment]
> > if the item's master category does not have a record in the products to categories table, we do not allow admins to find it. why?
> 
> I can't answer that, but it has flagged up this error, which is good. I've also had p2c/mcat issues in the past and now have a permanent check in place in the admin for products with an invalid mcat.

Steve, has the new p2c page helped with resolving those?

**Comment by torvista** [comment]
I think the issues I had was missing p2c entries for the mcat. Long time ago, and a long time since I saw an error flagged.
I would expect it was a mod that caused them in the first place (maybe Apsona).
So, certainly not core code/a recurring problem. So since there is no current issues, can't say if the new p2c is better, since I don't think it was broke in the first place in this respect.

**Comment by proseLA** [comment]
in looking at my code, i have not changed the SQL command with regards to this issue.  i think i did something that allowed clients to see where P2C records did not exist for master categories.  its a check after the fact.

**Comment by drbyte** [comment]
> I would expect it was a mod that caused them in the first place (maybe Apsona).

Ya, I was wondering if Apsona caused it. I've seen it on sites that used Apsona in the past.

> So since there is no current issues, can't say if the new p2c is better, since I don't think it was broke in the first place in this respect.

Fair. I was thinking that since the P2C page allows for moving/reassigning that maybe it may be a helpful tool in resolving any mismatches found. 
I'll try to set up a test to explore that tomorrow.

**Comment by proseLA** [comment]
i am looking at the new tool.  is there an easy place to just say, show me all products where there is a no p2c record for the products master category?  and then 1 click it creates them?

that has seemed to be the problem for me in the past, as well as the problem from the board.

**Comment by drbyte** [comment]
> i am looking at the new tool. is there an easy place to just say, show me all products where there is a no p2c record for the products master category? and then 1 click it creates them?

No, we don't have a report of "all products w/o a matching p2c record". I have a bespoke query for it, but am reluctant to build a dedicated page just for that: it seems a wrong reason for a dedicated page.

I was more interested in knowing that if you went to the p2c page for a product that doesn't have the match (say by setting the URL param manually), would it easily let you fix it there?

**Comment by scottcwilson** [comment]
If you use Audit, it finds things like this.

https://www.zen-cart.com/downloads.php?do=file&id=2261

**Comment by proseLA** [comment]
> If you use Audit, it finds things like this.
> 
> https://www.zen-cart.com/downloads.php?do=file&id=2261

i am as much a fan as the next guy of building a system to monitor the system; but i'm not really interesting in looking at a plugin to address an issue that i think can safely be addressed by removing the P2C table from the join in the products query.

> > i am looking at the new tool. is there an easy place to just say, show me all products where there is a no p2c record for the products master category? and then 1 click it creates them?
> 
> No, we don't have a report of "all products w/o a matching p2c record". I have a bespoke query for it, but am reluctant to build a dedicated page just for that: it seems a wrong reason for a dedicated page.

ok.
> 
> I was more interested in knowing that if you went to the p2c page for a product that doesn't have the match (say by setting the URL param manually), would it easily let you fix it there?

you have to know the product has a problem to start off with.  now mind you, clients will call and say there is a problem as this item is not appearing on the search, and i will track it down, and then attempt to find all other products who have a similar problem.

as far as this p2c page, 10 minutes was the first time i looked at it...

**Comment by drbyte** [comment]
> i think can safely be addressed by removing the P2C table from the join in the products query

Yes, it can. That's why I marked this as a Confirmed Bug.

> you have to know the product has a problem to start off with. 

Of course.

The point of my question was to determine whether there's already a built-in tool we can use to fix it.

Merely updating the query solves very little, because even if you can "find" the product you don't know there's a problem with the it, and even if you did, clicking Edit can't always be used to fix the problem because the product-edit page doesn't let you select categories unless it's linked to multiple categories.

My questions were around whether the tools that (the collective) y'all have been updating/rearranging/adding-to in the admin can already fix it.

**Comment by drbyte** [comment]
Unfortunately it looks like most of the admin editing-related pages that do any checks whether to display master_categories_id for editing only check "is the product linked to more than one category".

There are a couple built-in functions that can be leveraged for detection in real-time, depending on the need.

I'm debating whether the "move product to category" sidebar might be the right place to offer a selection. The dropdown could be massive though. Granted, there's already a massive dropdown in the search at the top, but it's unrelated.
The P2C or products_price_manager screen could be used too, as could the product-edit screen. All would require some form of alteration to be allowed to pick a specific master category, instead of only choosing from already-linked-categories.

**Comment by scottcwilson** [comment]
Even before this change if you had a product and you managed to clobber the P2C entry, search would fail to find it. (I have a client who reliably manages to do this, which is why I built audit.).

**Comment by torvista** [comment]
My permanent check for the missing mcat in p2c table runs in the admin header and I note no overhead with 5k products.
I even constructed the sql in the message to run in the tool.

With hindsight, I don't see why it cannot auto-add the p2c entry and just have a subsequent warning message and generate a debug that this correction was required....for further investigation

Otherwise , yes P2C change of master category will correct this. But its a bit of a non-intuitive faff to "move" the product about to correct it.

Other option is adding the function (to add missing p2c entries)  to Store Manager. I see this as a similar "correction for something that should not happen" as reset price sorter.

**Comment by torvista** [comment]
To clarify my previous post was referring to a product that did not have a matching p2c entry for it's mcat.
That is assuming the product mcat is correct.

But the issue here is that the product mcat is wrong on a duplication to another category. 

I find this does not occur with the ZC156 version of copy_product_confirm as the insert is explicit/uses the passed category id, not the original product mcat id
`'" . zen_db_input($categories_id) . "')");`

**Comment by drbyte** [comment]
> In my test, duplicating a product from catA to catB:
> 
> new product has master category of catA as it uses the same as source product - incorrectly.
> 
> new product has p2c category of catB - correct, as per product_copy_confirm.php:
> 
> ```
>         $db->Execute("INSERT INTO " . TABLE_PRODUCTS_TO_CATEGORIES . " (products_id, categories_id)
>                       VALUES (" . $dup_products_id . ", " . $categories_id . ")");
> ```
> 
> I think an extra line after this is the easiest fix, no?

> But the issue here is that the product mcat is wrong on a duplication to another category.
> 
> I find this does not occur with the ZC156 version of copy_product_confirm as the insert is explicit/uses the passed category id, not the original product mcat id
> `'" . zen_db_input($categories_id) . "')");`

It was a bug created in 1.5.7 when streamlining the copying of fields. 
Fixed via https://github.com/zencart/zencart/commit/79ee29d693eb7a9ef2b99501ef1866ac982e7aa3

**Comment by drbyte** [comment]
Turns out the p2c join is crucial in order to both display cross-linked products and to filter things to only show products in current category when browsing without search.

This is fixed by #3787
